File: app/src/main/java/com/antony/muzei/pixiv/ui/fragments/MainPreferenceFragment.java
/*
 *     This file is part of PixivforMuzei3.
 *
 *     PixivforMuzei3 is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program  is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.antony.muzei.pixiv.ui.fragments;

import android.app.Activity;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.os.Environment;
import android.util.Log;
import android.widget.Toast;

import androidx.preference.MultiSelectListPreference;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;
import androidx.preference.PreferenceManager;
import androidx.work.WorkManager;

import com.antony.muzei.pixiv.PixivArtProviderDefines;
import com.antony.muzei.pixiv.PixivArtWorker;
import com.antony.muzei.pixiv.R;
import com.antony.muzei.pixiv.ui.activity.LoginActivity;

import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

public class MainPreferenceFragment extends PreferenceFragmentCompat
{
	private static final int REQUEST_CODE_LOGIN = 1;
	private String oldUpdateMode, newUpdateMode;
	private String oldTag, newTag;
	private String oldArtist, newArtist;

	@Override
	public void onCreate(Bundle savedInstanceState)
	{
		super.onCreate(savedInstanceState);
		addPreferencesFromResource(R.xml.main_preference_layout);

		// If Muzei is not installed, this will redirect the user to Muzei's Play Store listing
		if (!isMuzeiInstalled())
		{
			final String appPackageName = "net.nurik.roman.muzei"; // getPackageName() from Context or Activity object
			try
			{
				startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=" + appPackageName)));
			} catch (android.content.ActivityNotFoundException ex)
			{
				startActivity(new Intent(Intent.ACTION_VIEW,
						Uri.parse("https://play.google.com/store/apps/details?id=" + appPackageName)));
			}
		}

		SharedPreferences sharedPrefs = PreferenceManager.getDefaultSharedPreferences(getContext());

		// Stores user toggleable variables into a temporary store for later comparison in onStop()
		// If the value of the preference on Activity creation is different to Activity stop, then take certain action
		oldUpdateMode = sharedPrefs.getString("pref_updateMode", "");
		newUpdateMode = oldUpdateMode;

		oldTag = sharedPrefs.getString("pref_tagSearch", "");
		newTag = oldTag;

		oldArtist = sharedPrefs.getString("pref_artistId", "");
		newArtist = oldArtist;

		// Ensures that the user has logged in first before selecting any update mode requiring authentication
		// Reveals UI elements as needed depending on Update Mode selection
		findPreference("pref_updateMode").setOnPreferenceChangeListener((preference, newValue) ->
		{
			boolean isAuthUpdateMode = Arrays.asList(PixivArtProviderDefines.AUTH_MODES)
					.contains(newValue.toString());
			// User has selected an authenticated feed mode, but has not yet logged in as evidenced
			// by the lack of an access token
			if (isAuthUpdateMode && sharedPrefs.getString("accessToken", "").isEmpty())
			{
				Toast.makeText(getContext(), getString(R.string.toast_loginFirst), Toast.LENGTH_SHORT).show();
				return false;
			}
			// If any of the auth feed modes, reveal login Preference Category, reveal the auth NSFW filtering,
			// and hide the ranking NSFW filtering
			boolean authFeedModeSelected = Arrays.asList(PixivArtProviderDefines.AUTH_MODES)
					.contains(newValue);

			findPreference("pref_authFilterSelect").setVisible(authFeedModeSelected);
			findPreference("pref_rankingFilterSelect").setVisible(!authFeedModeSelected);

			findPreference("pref_tagSearch").setVisible(newValue.equals("tag_search"));
			findPreference("pref_artistId").setVisible(newValue.equals("artist"));
			return true;
		});

		// All this is needed for the arbitrary selection NSFW filtering
		// Resets to default SFW filtering is all options are unchecked
		// Prints a summary string based on selection
		// Updates authFilterSelectPref summary as user updates it
		MultiSelectListPreference authFilterSelectPref = findPreference("pref_authFilterSelect");
		authFilterSelectPref.setOnPreferenceChangeListener((preference, newValue) ->
		{
			// Resets to SFW on empty selection
			// for some reason a length of 2 is an empty selection
			if (newValue.toString().length() == 2)
			{
				Set<String> defaultSet = new HashSet<>();
				defaultSet.add("2");
				authFilterSelectPref.setValues(defaultSet);

				SharedPreferences.Editor editor = sharedPrefs.edit();
				editor.putStringSet("pref_authFilterSelect", defaultSet);
				editor.commit();
				authFilterSelectPref.setSummary("SFW");
				return false;
			}

			// Prints a comma delimited string of user selections. There is no trailing comma
			// TODO there's gotta be a better wau of doing this
			String str = newValue.toString();
			ArrayList<Integer> arrayList = new ArrayList<>();
			for (int i = 0; i < str.length(); i++)
			{
				if (Character.isDigit(str.charAt(i)))
				{
					arrayList.add(Character.getNumericValue(str.charAt(i)));
				}
			}
			String[] authFilterEntriesPossible = getResources().getStringArray(R.array.pref_authFilterLevel_entries);
			StringBuilder stringBuilderAuth = new StringBuilder();
			for (int i = 0; i < arrayList.size(); i++)
			{
				stringBuilderAuth.append(authFilterEntriesPossible[(arrayList.get(i) - 2) / 2]);
				if (i != arrayList.size() - 1)
				{
					stringBuilderAuth.append(", ");
				}
			}
			String summaryAuth = stringBuilderAuth.toString();

			authFilterSelectPref.setSummary(summaryAuth);

			return true;
		});

		// updates ranking nsfw select summary on preference change
		MultiSelectListPreference rankingFilterSelectPref = findPreference("pref_rankingFilterSelect");
		rankingFilterSelectPref.setOnPreferenceChangeListener((preference, newValue) ->
		{
			// for some reason 2 is an empty selection
			if (newValue.toString().length() == 2)
			{
				Log.v("MANUAL", "pref change empty set");
				Set<String> defaultSet = new HashSet<>();
				defaultSet.add("0");
				rankingFilterSelectPref.setValues(defaultSet);

				SharedPreferences.Editor editor = sharedPrefs.edit();
				editor.putStringSet("pref_rankingFilterSelect", defaultSet);
				editor.commit();
				rankingFilterSelectPref.setSummary("SFW");
				return false;
			}

			String str = newValue.toString();
			ArrayList<Integer> arrayList = new ArrayList<>();
			for (int i = 0; i < str.length(); i++)
			{
				if (Character.isDigit(str.charAt(i)))
				{
					arrayList.add(Character.getNumericValue(str.charAt(i)));
				}
			}
			String[] rankingEntriesAvailable = getResources().getStringArray(R.array.pref_rankingFilterLevel_entries);
			StringBuilder stringBuilderRanking = new StringBuilder();
			for (int i = 0; i < arrayList.size(); i++)
			{
				stringBuilderRanking.append(rankingEntriesAvailable[arrayList.get(i)]);
				if (i != arrayList.size() - 1)
				{
					stringBuilderRanking.append(", ");
				}
			}
			String summaryRanking = stringBuilderRanking.toString();

			rankingFilterSelectPref.setSummary(summaryRanking);

			return true;
		});

		// Generates the ranking NSFW filter summary during activity startup
		Set<String> chosenLevelsSetRanking = sharedPrefs.getStringSet("pref_rankingFilterSelect", null);
		String[] chosenLevelsRanking = chosenLevelsSetRanking.toArray(new String[0]);
		String[] entriesAvailableRanking = getResources().getStringArray(R.array.pref_rankingFilterLevel_entries);
		StringBuilder stringBuilderRanking = new StringBuilder();
		for (int i = 0; i < chosenLevelsRanking.length; i++)
		{
			stringBuilderRanking.append(entriesAvailableRanking[Integer.parseInt(chosenLevelsRanking[i])]);
			if (i != chosenLevelsRanking.length - 1)
			{
				stringBuilderRanking.append(", ");
			}
		}
		String summaryRanking = stringBuilderRanking.toString();
		rankingFilterSelectPref.setSummary(summaryRanking);

		// Generates the authFilterSelectPref summary during activity startup
		// TODO combine this with the above summary setting code section
		Set<String> chosenLevelsSet = sharedPrefs.getStringSet("pref_authFilterSelect", null);
		String[] chosenLevels = chosenLevelsSet.toArray(new String[0]);
		String[] entriesAvailableAuth = getResources().getStringArray(R.array.pref_authFilterLevel_entries);
		StringBuilder stringBuilderAuth = new StringBuilder();
		for (int i = 0; i < chosenLevels.length; i++)
		{
			stringBuilderAuth.append(entriesAvailableAuth[(Integer.parseInt(chosenLevels[i]) - 2) / 2]);
			if (i != chosenLevels.length - 1)
			{
				stringBuilderAuth.append(", ");
			}
		}
		String summaryAuth = stringBuilderAuth.toString();
		authFilterSelectPref.setSummary(summaryAuth);

		// Reveal the tag_search or artist_id EditTextPreference and write the summary if update mode matches
		String updateMode = sharedPrefs.getString("pref_updateMode", "daily");
		if (Arrays.asList(PixivArtProviderDefines.AUTH_MODES)
				.contains(updateMode))
		{
			findPreference("pref_authFilterSelect").setVisible(true);
			findPreference("prefCat_loginSettings").setVisible(true);
			if (updateMode.equals("tag_search"))
			{
				Preference tagSearch = findPreference("pref_tagSearch");
				tagSearch.setVisible(true);
				tagSearch.setSummary(sharedPrefs.getString("pref_tagSearch", ""));
			} else if (updateMode.equals("artist"))
			{
				Preference artistId = findPreference("pref_artistId");
				artistId.setVisible(true);
				artistId.setSummary(sharedPrefs.getString("pref_artistId", ""));
			}
		} else
		{
			findPreference("pref_rankingFilterSelect").setVisible(true);
		}

		// Preference that immediately clears Muzei's image cache when pressed
		findPreference(getString(R.string.button_clearCache)).setOnPreferenceClickListener(preference ->
		{
			WorkManager.getInstance().cancelUniqueWork("ANTONY");
			File dir = getContext().getExternalFilesDir(Environment.DIRECTORY_PICTURES);

			String[] children = dir.list();
			for (String child : children)
			{
				new File(dir, child).delete();
			}
			PixivArtWorker.enqueueLoad(true);
			Toast.makeText(getContext(), getString(R.string.toast_clearingCache), Toast.LENGTH_SHORT).show();
			return true;
		});

		// On app launch set the Preference to show to appropriate text if logged in
		Preference loginActivityPreference = findPreference("pref_login");
		if (sharedPrefs.getString("accessToken", "").isEmpty())
		{
			loginActivityPreference.setTitle(R.string.prefTitle_loginButton);
			loginActivityPreference.setSummary(R.string.prefSummary_notLoggedIn);
		} else
		{
			loginActivityPreference.setTitle(R.string.prefTitle_logoutButton);
			loginActivityPreference.setSummary(getString(R.string.prefSummary_LoggedIn) + " " + sharedPrefs.getString("name", ""));
		}
		// Users click this preference to execute the login or logout
		loginActivityPreference.setOnPreferenceClickListener(preference ->
		{
			if (sharedPrefs.getString("accessToken", "").isEmpty())
			{
				Intent intent = new Intent(getContext(), LoginActivity.class);
				startActivityForResult(intent, REQUEST_CODE_LOGIN);
			} else
			{
				// Logging out
				SharedPreferences.Editor editor = sharedPrefs.edit();
				editor.remove("accessTokenIssueTime");
				editor.remove("name");
				editor.remove("accessToken");
				editor.remove("userId");
				editor.remove("refreshToken");

				loginActivityPreference.setTitle(R.string.prefTitle_loginButton);
				loginActivityPreference.setSummary(R.string.prefSummary_notLoggedIn);
				// If the user has an authenticated feed mode, reset it to daily ranking on logout
				if (Arrays.asList(PixivArtProviderDefines.AUTH_MODES)
						.contains(updateMode))
				{
					editor.putString("pref_updateMode", "daily");
				}
				editor.commit();
			}
			return true;
		});
	}

	@Override
	public void onActivityResult(int requestCode, int resultCode, Intent data)
	{
		super.onActivityResult(requestCode, resultCode, data);
		// Currently only used by LoginActivity
		if (requestCode == REQUEST_CODE_LOGIN && resultCode == Activity.RESULT_OK)
		{
			Preference loginButtonMain = findPreference("pref_login");

			loginButtonMain.setSummary(R.string.prefSummary_LoggedIn + data.getStringExtra("username"));
			loginButtonMain.setTitle(R.string.prefTitle_logoutButton);
		}
	}

	@Override
	public void onCreatePreferences(Bundle savedInstanceState, String rootKey)
	{
	}

	// Functions in here action only on app exit
	@Override
	public void onStop()
	{
		super.onStop();

		SharedPreferences sharedPrefs = PreferenceManager.getDefaultSharedPreferences(getContext());
		newUpdateMode = sharedPrefs.getString("pref_updateMode", "");
		newTag = sharedPrefs.getString("pref_tagSearch", "");
		newArtist = sharedPrefs.getString("pref_artistId", "");

		// If user has changed update, filter mode, or search tag:
		// Immediately stop any pending work, clear the Provider of any Artwork, and then toast
		if (!oldUpdateMode.equals(newUpdateMode) || !oldTag.equals(newTag)
				|| !oldArtist.equals(newArtist))
		{
			WorkManager.getInstance().cancelUniqueWork("ANTONY");
			File dir = getContext().getExternalFilesDir(Environment.DIRECTORY_PICTURES);
			String[] children = dir.list();
			for (String child : children)
			{
				new File(dir, child).delete();
			}
			PixivArtWorker.enqueueLoad(true);
			if (!oldUpdateMode.equals(newUpdateMode))
			{
				Toast.makeText(getContext(), getString(R.string.toast_newUpdateMode), Toast.LENGTH_SHORT).show();
			} else if (!oldArtist.equals(newArtist))
			{
				Toast.makeText(getContext(), getString(R.string.toast_newArtist), Toast.LENGTH_SHORT).show();
			} else if (!oldTag.equals(newTag))
			{
				Toast.makeText(getContext(), getString(R.string.toast_newTag), Toast.LENGTH_SHORT).show();
			} else
			{
				Toast.makeText(getContext(), getString(R.string.toast_newFilterSelect), Toast.LENGTH_SHORT).show();
			}
		}
	}

	// Redirects the user to Muzei's Play Store listing if it's not detected to be installed
	// TODO have a nicer dialog that explains why Muzei needs to be installed, instead of forcing them to the store
	private boolean isMuzeiInstalled()
	{
		boolean found = true;
		try
		{
			getContext().getPackageManager().getPackageInfo("net.nurik.roman.muzei", 0);
		} catch (PackageManager.NameNotFoundException ex)
		{
			found = false;
		}
		return found;
	}
}

File: app/src/main/java/com/antony/muzei/pixiv/settings/fragments/MainPreferenceFragment.kt
/*
 *     This file is part of PixivforMuzei3.
 *
 *     PixivforMuzei3 is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program  is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.antony.muzei.pixiv.settings.fragments

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.os.Environment
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.preference.*
import androidx.work.WorkManager
import com.antony.muzei.pixiv.PixivProviderConst.AUTH_MODES
import com.antony.muzei.pixiv.PixivProviderConst.PREFERENCE_PIXIV_ACCESS_TOKEN
import com.antony.muzei.pixiv.R
import com.antony.muzei.pixiv.login.LoginActivityWebview
import com.antony.muzei.pixiv.provider.PixivArtWorker.Companion.enqueueLoad
import com.google.android.material.snackbar.Snackbar
import java.util.*


class MainPreferenceFragment : PreferenceFragmentCompat() {
    private lateinit var oldUpdateMode: String
    private lateinit var newUpdateMode: String
    private lateinit var oldTag: String
    private lateinit var newTag: String
    private lateinit var oldArtist: String
    private lateinit var newArtist: String

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        setPreferencesFromResource(R.xml.main_preference_layout, rootKey)

        val sharedPrefs = PreferenceManager.getDefaultSharedPreferences(context)

        // Stores user toggleable variables into a temporary store for later comparison in onStop()
        // If the value of the preference on Activity creation is different to Activity stop, then take certain action
        oldUpdateMode = sharedPrefs.getString("pref_updateMode", "daily") ?: "daily"
        oldTag = sharedPrefs.getString("pref_tagSearch", "") ?: ""
        oldArtist = sharedPrefs.getString("pref_artistId", "") ?: ""

        // Ensures that the user has logged in first before selecting any update mode requiring authentication
        // Reveals UI elements as needed depending on Update Mode selection
        val updateModePref = findPreference<DropDownPreference>("pref_updateMode")
        updateModePref?.onPreferenceChangeListener =
            Preference.OnPreferenceChangeListener setOnPreferenceChangeListener@{ _: Preference?, newValue: Any ->
                // User has selected an authenticated feed mode, but has not yet logged in as evidenced
                // by the lack of an access token
                if (AUTH_MODES.contains(newValue) && sharedPrefs.getString(
                        PREFERENCE_PIXIV_ACCESS_TOKEN,
                        ""
                    )!!.isEmpty()
                ) {
                    Snackbar.make(
                        requireView(), R.string.toast_loginFirst,
                        Snackbar.LENGTH_SHORT
                    ).show()
                    return@setOnPreferenceChangeListener false
                }
                // If any of the auth feed modes, reveal login Preference Category, reveal the auth NSFW filtering,
                // and hide the ranking NSFW filtering
                val authFeedModeSelected = AUTH_MODES.contains(newValue)
                findPreference<Preference>("pref_authFilterSelect")?.isVisible =
                    authFeedModeSelected
                findPreference<Preference>("pref_rankingFilterSelect")?.isVisible =
                    !authFeedModeSelected
                findPreference<Preference>("pref_tagSearch")?.isVisible = newValue == "tag_search"
                findPreference<Preference>("pref_tagLanguage")?.isVisible = newValue == "tag_search"
                findPreference<Preference>("pref_artistId")?.isVisible = newValue == "artist"
                true
            }

        // All this is needed for the arbitrary selection NSFW filtering
        // Will default to only SFW if no filtering modes are selected
        // Prints a summary string based on selection
        // Updates authFilterSelectPref summary as user updates it
        // SimpleSummaryProvider does not support MultiSelectListPreference
        findPreference<MultiSelectListPreference>("pref_authFilterSelect")?.let {
            it.onPreferenceChangeListener =
                Preference.OnPreferenceChangeListener setOnPreferenceChangeListener@{ _: Preference?, newValue: Any ->
                    // Reset to SFW on empty selection
                    @Suppress("UNCHECKED_CAST")
                    if ((newValue as Set<String>).isEmpty()) {
                        it.values = setOf("2")
                        sharedPrefs.edit().apply {
                            putStringSet("pref_authFilterSelect", setOf("2"))
                            apply()
                        }
                        it.summary = "SFW"
                        return@setOnPreferenceChangeListener false
                    }
                    it.summary =
                        authSummaryStringGenerator(newValue)

                    true
                }
            it.summary = authSummaryStringGenerator(
                sharedPrefs.getStringSet(
                    "pref_authFilterSelect",
                    setOf("2")
                ) as Set<String>
            )
        }

        // Updates ranking SFW filtering preference
        // Same manner as above, the auth modes
        findPreference<MultiSelectListPreference>("pref_rankingFilterSelect")?.let {
            it.onPreferenceChangeListener =
                Preference.OnPreferenceChangeListener setOnPreferenceChangeListener@{ _: Preference?, newValue: Any ->
                    // Reset to SFW on empty selection
                    @Suppress("UNCHECKED_CAST")
                    if ((newValue as Set<String>).isEmpty()) {
                        it.values = setOf("0")
                        sharedPrefs.edit().apply {
                            putStringSet("pref_rankingFilterSelect", setOf("0"))
                            apply()
                        }
                        it.summary = "SFW"
                        return@setOnPreferenceChangeListener false
                    }

                    it.summary =
                        rankingSummaryStringGenerator(newValue)

                    true
                }
            // Generates the ranking NSFW filter summary during activity startup
            it.summary =
                rankingSummaryStringGenerator(
                    sharedPrefs.getStringSet("pref_rankingFilterSelect", setOf("0")) as Set<String>
                )
        }

        // Reveal the tag_search or artist_id EditTextPreference and write the summary if update mode matches
        val updateMode = sharedPrefs.getString("pref_updateMode", "daily")
        if (AUTH_MODES.contains(updateMode)) {
            findPreference<Preference>("pref_authFilterSelect")?.isVisible = true
            findPreference<Preference>("prefCat_loginSettings")?.isVisible = true
            if (updateMode == "tag_search") {
                findPreference<Preference>("pref_tagSearch")?.let {
                    it.isVisible = true
                }
                findPreference<Preference>("pref_tagLanguage")?.let {
                    it.isVisible = true
                }
            } else if (updateMode == "artist") {
                findPreference<Preference>("pref_artistId")?.let {
                    it.isVisible = true
                }
            }
        } else {
            findPreference<Preference>("pref_rankingFilterSelect")?.isVisible = true
        }

        // Preference that immediately clears Muzei's image cache when pressed
        findPreference<Preference>(getString(R.string.button_clearCache))?.onPreferenceClickListener =
            Preference.OnPreferenceClickListener {
                WorkManager.getInstance(requireContext()).cancelUniqueWork("ANTONY")
                requireContext().getExternalFilesDir(Environment.DIRECTORY_PICTURES)
                    ?.deleteRecursively()
                enqueueLoad(true, context)
                Snackbar.make(
                    requireView(), R.string.toast_clearingCache,
                    Snackbar.LENGTH_SHORT
                )
                    .show()
                newUpdateMode = oldUpdateMode
                true
            }

        val loginActivityLauncher =
            registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
                if (result.resultCode == Activity.RESULT_OK) {
                    val data = result.data
                    val loginButtonMain = findPreference<Preference>("pref_login")
                    loginButtonMain!!.summary =
                        getString(R.string.prefSummary_LoggedIn) + " " + data!!.getStringExtra("username")
                    loginButtonMain.title = getString(R.string.prefTitle_logoutButton)
                }
            }

        findPreference<Preference>("pref_login")?.let {
            // On app launch set the Preference to show to appropriate text if logged in
            if (sharedPrefs.getString(PREFERENCE_PIXIV_ACCESS_TOKEN, "")?.isEmpty() == true) {
                it.title = getString(R.string.prefTitle_loginButton)
                it.summary = getString(R.string.prefSummary_notLoggedIn)
            } else {
                it.title = getString(R.string.prefTitle_logoutButton)
                it.summary =
                    "${getString(R.string.prefSummary_LoggedIn)} ${
                        sharedPrefs.getString(
                            "name",
                            ""
                        )
                    }"
            }

            // Users click this preference to execute the login or logout
            it.onPreferenceClickListener = Preference.OnPreferenceClickListener {
                loginActivityLauncher.launch(Intent(activity, LoginActivityWebview::class.java))
                true
            }
        }
    }

    // 0 or 1 correspond to SFW or NSFW respectively
    private fun rankingSummaryStringGenerator(selection: Set<String>): String {
        val rankingFilterEntriesPossible =
            resources.getStringArray(R.array.pref_rankingFilterLevel_entries)
        return StringBuilder().let {
            selection.forEachIndexed { _, element ->
                it.append(rankingFilterEntriesPossible[element.toInt()])
                it.append(", ")
            }
            it.setLength(it.length - 2)
            it.toString()
        }
    }

    // Returns a comma delimited string of user selections. There is no trailing comma
    // newValue is a HashSet that can contain 2, 4, 6, or 8, and corresponds to
    // SFW, Slightly Ecchi, Fairly Ecchi, and R18 respectively
    // 2, 4, 6, 8 was selected to match with the values used in Pixiv JSON
    private fun authSummaryStringGenerator(selection: Set<String>): String {
        val authFilterEntriesPossible =
            resources.getStringArray(R.array.pref_authFilterLevel_entries)
        return StringBuilder().let {
            selection.forEachIndexed { _, element ->
                // Translation from {2, 4, 6, 8} to {0, 1, 2, 3}
                it.append(authFilterEntriesPossible[(element.toInt() - 2) / 2])
                it.append(", ")
            }
            it.setLength(it.length - 2)
            it.toString()
        }
    }

    // Functions in here action only on app exit
    override fun onStop() {
        super.onStop()
        val sharedPrefs = PreferenceManager.getDefaultSharedPreferences(requireContext())
        newUpdateMode = sharedPrefs.getString("pref_updateMode", "") ?: ""
        newTag = sharedPrefs.getString("pref_tagSearch", "") ?: ""
        newArtist = sharedPrefs.getString("pref_artistId", "") ?: ""

        // If user has changed update, filter mode, or search tag:
        // Immediately stop any pending work, clear the Provider of any Artwork, and then toast
        if (oldUpdateMode != newUpdateMode || oldTag != newTag
            || oldArtist != newArtist
        ) {
            WorkManager.getInstance(requireContext()).cancelUniqueWork("ANTONY")
            requireContext().getExternalFilesDir(Environment.DIRECTORY_PICTURES)
                ?.deleteRecursively()
            enqueueLoad(true, context)
            if (oldUpdateMode != newUpdateMode) {
                Toast.makeText(context, getString(R.string.toast_newUpdateMode), Toast.LENGTH_SHORT)
                    .show()
            } else if (oldArtist != newArtist) {
                Toast.makeText(context, getString(R.string.toast_newArtist), Toast.LENGTH_SHORT)
                    .show()
            } else if (oldTag != newTag) {
                Toast.makeText(context, getString(R.string.toast_newTag), Toast.LENGTH_SHORT).show()
            } else {
                Toast.makeText(
                    context,
                    getString(R.string.toast_newFilterSelect),
                    Toast.LENGTH_SHORT
                ).show()
            }
        }
    }
}
